docs/spec.md

```diff
@@ -5789,10 +5781,12 @@ A StableHLO program is executed by providing input values to the `main` function
 and computing output values. Output values of a function are computed by
 executing the graph of ops rooted in the corresponding `return` op.
 
-The execution order is implementation-defined, as long as ops are executed
-before their uses. Possible execution orders of the example program above are
-`%0` → `%1` → `%2` → `%3` → `%4` → `return` or `%3` → `%0` → `%1` → `%2` → `%4`
-→ `return`.
+The execution order is implementation-defined as long as it is aligned with
+dataflow, i.e. if ops are executed before their uses. In StableHLO, all
+side-effecting ops consume and/or produce token(s) to reify the side effects as
+a value that other operations can take a data dependency on. Possible execution
```

Comment:
I'd like to propose a reformulation: "In StableHLO, all side-effecting ops consume a token and produce a token (multiple tokens can be multiplexed via `after_all`), so the execution order of side effects is also aligned with dataflow":
  1) These ops don't just consume and/or produce an unspecified number of tokens. Each such op consumes exactly one and produces exactly one token.
  2) The second part of rewording explicitly focuses on the execution order since this is what this paragraph is talking about.

Reply:
1. Thanks, I originally misinterpreted "variadic number of tensors or tokens" in Case, If, and While op.
2. Sounds good to me.